tests/core_tests/tx_validation.cpp

```diff
@@ -2108,3 +2108,533 @@ bool tx_pool_semantic_validation::generate(std::vector<test_event_entry>& events
+
+      {
+        {
+          crypto::point_t point_key_image{};
+
+          point_key_image.from_string("bc2ddcc593039f0ece76eeefd91c2c3e8c4aca879b6e3adaaf0c9288da88c0f0");
+          input.k_image = point_key_image.to_key_image();
+        }
+
+        // A container is selected by an amount specified in an input. ZC outputs have .amount equals to 0. Thus, the input has .amount equals to 0.
+        input.amount = 0;
+        tx_1.vin.push_back(input);
+        CHECK_AND_ASSERT_EQ(tx_1.vin.size(), 1);
+      }
+    }
+
+    tx_1.vout.push_back(tx_out_zarcanum{});
+    tx_1.vout.push_back(tx_out_zarcanum{});
+    CHECK_AND_ASSERT_EQ(tx_1.vout.size(), 2);
+    tx_1.extra.push_back(zarcanum_tx_data_v1{TESTS_DEFAULT_FEE});
+    CHECK_AND_ASSERT_EQ(tx_1.extra.size(), 1);
+    DO_CALLBACK(events, "mark_invalid_tx");
+    ADD_CUSTOM_EVENT(events, tx_1);
+    DO_CALLBACK_PARAMS_STR(events, "assert_to_key_input_refers_zarcanum_output_is_wrong", t_serializable_object_to_blob(tx_1));
+  }
+
+  // An input of the type "txin_zc_input" refers by a "ref_by_id" object to an output with a target of the type "txout_to_key".
+  {
+    DO_CALLBACK_PARAMS(events, "check_hardfork_active", size_t{ZANO_HARDFORK_04_ZARCANUM});
+    MAKE_TX_FEE(events, tx_0, miner, miner, MK_TEST_COINS(2), TESTS_DEFAULT_FEE, top);
+
+    {
+      MAKE_NEXT_BLOCK_TX1(events, blk, top, miner, tx_0);
+      top = blk;
+    }
+
+    {
+      REWIND_BLOCKS(events, blk_r, top, miner);
+      top = blk_r;
+    }
+
+    transaction tx_1{};
+
+    tx_1.version = 2;
+
+    {
+      txin_zc_input input{};
+
+      input.key_offsets.emplace_back(ref_by_id{get_transaction_hash(tx_00), 0});
+
+      {
+        const auto& output{tx_00.vout.at(boost::get<ref_by_id>(input.key_offsets.front()).n)};
+
+        CHECK_AND_ASSERT_EQ(output.type(), typeid(tx_out_bare));
+        CHECK_AND_ASSERT_EQ(boost::get<tx_out_bare>(output).target.type(), typeid(txout_to_key));
+      }
+
+      CHECK_AND_ASSERT_EQ(input.key_offsets.size(), 1);
+      input.key_offsets.emplace_back(ref_by_id{get_transaction_hash(tx_01), 0});
+
+      {
+        const auto& output{tx_01.vout.at(boost::get<ref_by_id>(input.key_offsets.at(1)).n)};
+
+        CHECK_AND_ASSERT_EQ(output.type(), typeid(tx_out_zarcanum));
+      }
+
+      CHECK_AND_ASSERT_EQ(input.key_offsets.size(), 2);
+
+      {
+        crypto::point_t point_key_image{};
+
+        point_key_image.from_string("3131d0f71373ff2114e8174d1820122d8031d51182c037add27b8cf2ddd4349a");
+        input.k_image = point_key_image.to_key_image();
+      }
+
+      tx_1.vin.push_back(input);
+      CHECK_AND_ASSERT_EQ(tx_1.vin.size(), 1);
+    }
+
+    tx_1.vout.push_back(tx_out_zarcanum{});
+    tx_1.vout.push_back(tx_out_zarcanum{});
+    CHECK_AND_ASSERT_EQ(tx_1.vout.size(), 2);
+    tx_1.extra.push_back(zarcanum_tx_data_v1{TESTS_DEFAULT_FEE});
+    CHECK_AND_ASSERT_EQ(tx_1.extra.size(), 1);
+    DO_CALLBACK(events, "mark_invalid_tx");
```

Comment:
`mark_invalid_tx` marks the next event as an invalid tx. This makes not much sense since the next event is a callback.